### PR TITLE
Quest Tracker Feature

### PR DESCRIPTION
**Pull request by joseluis2g**
Adding quest tracker features, mostly all credits go to saiyansking, I  just updated this to work with Lua.
Also added a new value to quest.lua, missionId, this value should be unique and every mission needs one.
Added a new player event called onResetTrackedQuests, name tells you pretty much everything what it does.

### CODE REVIEW

File: data/lib/core/quests.lua [hunk 1]
```diff
@@ -831,29 +831,29 @@ if not Quests then
...
+						[6] = "You received Isimov\'s prayer: &quot;Unclean spirits shall be repelled&quot;. \z
 							Now, bring your stake to Amanda in Edron for the next line of the prayer.",
-						[7] = "You received Amanda's prayer: &quot;Wicked curses shall be broken&quot;. \z
+						[7] = "You received Amanda\'s prayer: &quot;Wicked curses shall be broken&quot;. \z
 							Now, bring your stake to Kasmir in Darashia for the next line of the prayer.",
-						[8] = "You received Kasmir's prayer: &quot;Let there be honour and humility&quot;. \z
+						[8] = "You received Kasmir\'s prayer: &quot;Let there be honour and humility&quot;. \z
 							Now, bring your stake to Rahkem in Ankrahmun for the next line of the prayer.",
-						[9] = "You received Rahkem's prayer: &quot;Let there be power and compassion&quot;. \z
+						[9] = "You received Rahkem\'s prayer: &quot;Let there be power and compassion&quot;. \z
```
Comment (gpedro) [inline data/lib/core/quests.lua:854]: why needs to escape the apostrophe?

File: data/lib/core/quests.lua [hunk 2]
```diff
@@ -5403,23 +5492,14 @@ function Player.getMissionDescription(self, questId, missionId)
...
-
 function Player.sendQuestLog(self)
 	local msg = NetworkMessage()
 	msg:addByte(0xF0)
 	msg:addU16(Game.getQuestsCount(self))
 	for questId = 1, #Quests do
 		if self:questIsStarted(questId) then
 			msg:addU16(questId)
-			msg:addString(Quests[questId].name .. "" .. (self:questIsCompleted(questId) and " (completed)" or ""))
+			msg:addString(Quests[questId].name .. ""..(self:questIsCompleted(questId) and " (completed)" or ""))
```
Comment (gpedro) [inline data/lib/core/quests.lua]: doesn't need two concat

File: data/lib/core/quests.lua [hunk 3]
```diff
@@ -5439,27 +5518,49 @@ function Player.sendQuestLine(self, questId)
...
-		LastQuestlogUpdate[guid] = currentFrameTime
-		if Game.sendQuestLogUpdatedMessage(value) then
-			self:sendTextMessage(MESSAGE_EVENT_ADVANCE, "Your questlog has been updated.")
+function Player.sendTrackedQuests(self, remainingQuests, missions)
+	local msg = NetworkMessage()
+	msg:addByte(0xD0)
+	msg:addByte(0x01)
+	msg:addByte(remainingQuests)
+	local count = #missions
+	msg:addByte(count)
```
Comment (gpedro) [inline data/lib/core/quests.lua]: you can use directly
```lua
msg:addByte(#missions)
```

File: data/lib/core/quests.lua [hunk 4]
```diff
@@ -5439,27 +5518,49 @@ function Player.sendQuestLine(self, questId)
...
+	for i = 1, count do
```
Comment (gpedro) [inline data/lib/core/quests.lua]: you can use loop-pairs instead

**Review by gpedro — CHANGES_REQUESTED** [review]
LGTM, minor changes

File: data/lib/core/quests.lua (hunk 1 above)
Comment (joseluis2g) [reply data/lib/core/quests.lua:854]: there was an error with the script I did to add mission ids and escape characters so I just did a check for all escape characters to add this "\"

File: data/lib/core/quests.lua (hunk 2 above)
Comment (joseluis2g) [reply data/lib/core/quests.lua]: my bad

File: data/lib/core/quests.lua (hunk 3 above)
Comment (joseluis2g) [reply data/lib/core/quests.lua]: thats a fact

**Comment by joseluis2g** [comment]
Remove events call, now it uses modules to read the message.
Fix gpedro conflicts.

File: src/protocolgame.h [hunk 5]
```diff
@@ -172,7 +172,7 @@ class ProtocolGame final : public Protocol
 		void parseOpenChannel(NetworkMessage& msg);
 		void parseOpenPrivateChannel(NetworkMessage& msg);
 		void parseCloseChannel(NetworkMessage& msg);
-
+
```
Comment (marcosvf132) [inline src/protocolgame.h]: Empty commit

Comment (joseluis2g) [reply src/protocolgame.h]: Thanks

File: data/modules/scripts/questtrack/questtrack.lua [hunk 6]
```diff
@@ -0,0 +1,10 @@
+function onRecvbyte(player, msg, byte)
+	if (byte == 0xD0) then
```
Comment (gpedro) [inline data/modules/scripts/questtrack/questtrack.lua]: unnecessary parenthesis :D

**Review by gpedro — APPROVED** [review]

**Comment by majestyotbr** [comment]
Tested, working. Thanks for your contribution.

**Pull request merged**